sdk/python/kfp/compiler/compiler.py

```diff
@@ -192,27 +200,54 @@ def _create_pipeline(
     def _write_pipeline_spec_json(
         self,
         pipeline_spec: pipeline_spec_pb2.PipelineSpec,
-        output_path: str,
+        package_path: str,
+    ) -> None:
+        """Writes pipeline spec into a JSON file.
+        Args:
+            pipeline_spec: IR pipeline spec.
+            ouput_path: The file path to be written.
+        """
+
+        warnings.warn(
+            ("Compiling pipline spec to JSON is deprecated and will be "
+             "removed in a future version. Please compile to a YAML file by "
+             "providing a file path with .yaml extension instead."),
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
+
+        json_text = json_format.MessageToJson(pipeline_spec, sort_keys=True)
+
+        with open(package_path, 'w') as json_file:
+            json_file.write(json_text)
+
+    def _write_pipeline_spec_yaml(
+        self,
+        pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+        package_path: str,
     ) -> None:
-        """Writes pipeline spec into a json file.
+        """Writes pipeline spec into a YAML file.
 
         Args:
             pipeline_spec: IR pipeline spec.
             ouput_path: The file path to be written.
 
         Raises:
-            ValueError: if the specified output path doesn't end with the
-                acceptable extention.
+            ValueError: if the specified output path doesn't end with a
+                .yaml extention.
         """
-        json_text = json_format.MessageToJson(pipeline_spec, sort_keys=True)
 
-        if output_path.endswith('.json'):
-            with open(output_path, 'w') as json_file:
-                json_file.write(json_text)
-        else:
+        if not package_path.endswith(('.yaml', ".yml")):
             raise ValueError(
-                'The output path {} should ends with ".json".'.format(
-                    output_path))
+                'The output path {} should end with ".yaml".'.format(
+                    package_path))
+
+        # sort keys=False throughout retains PipelineSpec's order
```

Comment:
Does it? Your yaml test snapshot files aren't in PipelineSpec's order.

Reply:
The issue was with how I updated the golden snapshots (I converted the JSON to YAML, instead of rerunning the .py files to get the golden snapshots). This is now updated.

Reply:
Okay, I guess they probably made some change in the library. Previously, my experience with `MessageToJson` is that `sort_keys=False` will result nondeterministic order. So I doubted you would get the proto definition order, since you still rely on `MessageToJson` as the first step.
Anyway, it looks good now.